### PR TITLE
Fix login as other user

### PR DESCRIPTION
**Pull request by mkangia**
https://manage.dimagi.com/default.asp?278446#1508216

There is a js error when you visit https://www.commcarehq.org/hq/admin/auth_as/ which happns [here](https://github.com/dimagi/commcare-hq/blob/9d031cc561c5c3a956654a16891ae08da75072c7/corehq/apps/reports/static/reports/js/tabular.js#L63) because there is no `data`. This is happening for XHR requests to urls like "https://i.kissmetrics.com/i.js?_=1530283757604"

tabular.js?version=9233b57:63 Uncaught TypeError: Cannot read property 'slug' of undefined
    at HTMLDocument.<anonymous> (tabular.js?version=9233b57:63)
    at HTMLDocument.dispatch (jquery.min.js?version=1055018:3)
    at HTMLDocument.q.handle (jquery.min.js?version=1055018:3)
    at Object.trigger (jquery.min.js?version=1055018:4)
    at A (jquery.min.js?version=1055018:4)
    at HTMLScriptElement.c (jquery.min.js?version=1055018:4)
    at HTMLScriptElement.dispatch (jquery.min.js?version=1055018:3)
    at HTMLScriptElement.q.handle (jquery.min.js?version=1055018:3)

I don't think that js block needs to run for this page. Looks like js_options is a requirement of using the base view `hqadmin/hqadmin_base_report.html` so just added that as empty and now it returns out of that block.

I am not sure if there is a better way to address this like change the template view or update that JS to run `renderPage(data.slug, data.report_table_js_options);` only when there is data but that might bring unwanted changes or ignore real errors.

### CODE REVIEW

**Comment by orangejenny** [comment]
The error is coming out of tabular.js, which is included because this page inherits from hqadmin_base_report...but this page isn't a report and doesn't use tabular data. Why not inherit from two_column.html or base_section.html instead?

It's also not clear to me that this fixes the issue. When I check out your branch, I still see the behavior Kai reported (nothing happens if you fill in email, 404 if you fill in both email and domain).

**Comment by mkangia** [comment]
Thank you @orangejenny 

Did few things:
1. Added `UserAdministration` view and corresponding template to extract things needed for its children views from `hqadmin_base_report`. Removed any reference to `report` in this template as well.
2. Made some improvements: Render error messages and removed changes to url which eventually resulted in removing the `authenticate_as` js since it didn't do anything else.

The one change i made which changes the functionality is allowing login as web user as well. It was not clear if it was intended to be kept out to. Don't think it would do a harm.

File: corehq/apps/hqadmin/views/users.py [hunk 1]
```diff
@@ -93,6 +93,11 @@ def post(self, request, *args, **kwargs):
             request.user.backend = 'django.contrib.auth.backends.ModelBackend'
             login(request, request.user)
             return HttpResponseRedirect('/')
+        all_errors = form.errors.pop('__all__', None)
```
Comment (orangejenny) [inline corehq/apps/hqadmin/views/users.py]: This is a nice bit of improvement.

File: corehq/apps/hqadmin/static/hqadmin/js/authenticate_as.js [hunk 2]
```diff
@@ -1,12 +0,0 @@
-/* globals hqDefine */
```
Comment (orangejenny) [inline corehq/apps/hqadmin/static/hqadmin/js/authenticate_as.js]: Sweet, always happy to see code deleted.

**Review by orangejenny — APPROVED** [review]

File: corehq/apps/hqadmin/forms.py [hunk 3]
```diff
@@ -51,8 +51,9 @@ def clean(self):
             if user is None:
                 raise forms.ValidationError("Cannot find user '{}'".format(username))
 
-        if not user.is_commcare_user():
-            raise forms.ValidationError("User '{}' is not a CommCareUser".format(username))
+        if domain or '.commcarehq.org' in username:
+            if not user.is_commcare_user():
+                raise forms.ValidationError("User '{}' is not a CommCareUser".format(username))
```
Comment (esoergel) [inline corehq/apps/hqadmin/forms.py]: I think this should use `CouchUser.get_by_username` if its expected to sometimes return `WebUser`s.

File: corehq/apps/hqadmin/tests/test_authenticate_as.py [hunk 4]
```diff
@@ -55,14 +55,6 @@ def test_invalid_data(self):
         form = AuthenticateAsForm(data)
         self.assertFalse(form.is_valid())
 
-    def test_no_login_as_other_webuser(self):
-        data = {
-            'username': 'awebuser'
-        }
-        form = AuthenticateAsForm(data)
-        self.assertFalse(form.is_valid())
-        self.assertTrue('not a CommCareUser' in str(form.errors))
```
Comment (esoergel) [inline corehq/apps/hqadmin/tests/test_authenticate_as.py]: You could turn this into a test validating that logging in as a web user works

File: corehq/apps/hqadmin/views/users.py [hunk 5]
```diff
@@ -89,10 +93,15 @@ def post(self, request, *args, **kwargs):
...
+        if all_errors:
+            messages.error(request, ','.join(all_errors))
+        if form.errors:
+            messages.error(request, form.errors)
```
Comment (esoergel) [inline corehq/apps/hqadmin/views/users.py]: Could you instead render the form errors in the html?  Django has good support for this kind of thing:
https://docs.djangoproject.com/en/1.11/topics/forms/#rendering-form-error-messages

I expect crispy_forms does too - doesn't it automatically render both field specific errors and `non_field_errors`?

File: corehq/ex-submodules/auditcare/views.py [hunk 6]
```diff
@@ -114,31 +114,31 @@ def model_instance_history(request, model_name, model_uuid, *args, **kwargs):
...
-    context=RequestContext(request)
     db = AccessAudit.get_db()
     vals = db.view('auditcare/model_actions_by_id', group=True, group_level=1).all()
     # do a dict comprehension here because we know all the keys in this reduce are unique
-    model_dict= dict((x['value'][0], x['value']) for x in vals)
-    context['model_dict']=model_dict
-    return render_to_response('auditcare/model_changes.html', context)
+    model_dict = dict((x['value'][0], x['value']) for x in vals)
+    context = {'model_dict': model_dict}
+    return render(request, 'auditcare/model_changes.html', context)
```
Comment (esoergel) [inline corehq/ex-submodules/auditcare/views.py]: wow, some old-school Django in this file

**Review by esoergel — APPROVED** [review]

File: corehq/apps/hqadmin/forms.py (hunk 3 above)
Comment (mkangia) [reply corehq/apps/hqadmin/forms.py]: reverted to allow login as a mobile user only, so not taking any action for this.

**Comment by mkangia** [comment]
thanks for finding that PR @esoergel .. 

That answered few things for me... so i have updated this. commit messages should be clear but just to sum up:
1. allow login only as a mobile worker and make it more clear
2. bring back the js file (sorry @orangejenny) for audit to work correctly. Just fixed that to pick the base/root url instead of current url which was causing the error as mentioned in the ticket.
3. this view and some others were actually just broken. Looks like we don't use this at all. So fixed those.
4. Also made some random improvements
5. going to add some comments around so that someone naive like me does not mess it up.

**Comment by orangejenny** [comment]
> bring back the js file

No worries, js has its place ;)

> going to add some comments around so that someone naive like me does not mess it up.

A+

File: corehq/apps/hqadmin/views/users.py (hunk 5 above)
Comment (mkangia) [reply corehq/apps/hqadmin/views/users.py]: Interesting... i tried rendering errors with fields but that didn't work. Going to not worry about it for this and leave it the way it is. Sorry.

File: corehq/ex-submodules/auditcare/views.py (hunk 6 above)
Comment (mkangia) [reply corehq/ex-submodules/auditcare/views.py]: tell me about it.. :)

File: corehq/apps/hqadmin/tests/test_authenticate_as.py (hunk 4 above)
Comment (mkangia) [reply corehq/apps/hqadmin/tests/test_authenticate_as.py]: reverted.

**Pull request merged**